File: python-sdk/src/forta_agent/labels_api.py
import requests
from .label import Label
from .utils import get_forta_api_headers, get_forta_api_url


def get_labels(dict):
    forta_api_url = get_forta_api_url()
    headers = get_forta_api_headers()
    query_options = LabelQueryOptions(dict)
    payload = query_options.get_query()

    response = requests.request(
        "POST", forta_api_url, json=payload, headers=headers)

    if response.status_code != 200:
        raise Exception(response.text)

    json_response = response.json()
    errors = json_response.get('errors')
    if errors is not None and len(errors) > 0:
        raise Exception(errors[0].get('message'))

    data = json_response.get('data')
    return LabelsResponse(data.get('labels'))


class LabelQueryOptions:
    def __init__(self, dict):
        self.entities = dict.get('entities')
        self.labels = dict.get('labels')
        self.sourceIds = dict.get('source_ids')
        self.entityType = dict.get('entity_type')
        self.state = dict.get('state')
        self.createdSince = dict.get('created_since')
        self.createdBefore = dict.get('created_before')
        self.first = dict.get('first')
        starting_cursor = dict.get('starting_cursor')
        if starting_cursor is not None:
            self.after = {
                'pageToken': starting_cursor.page_token if isinstance(starting_cursor, LabelCursor) else starting_cursor.get('page_token')
            }

    def get_query(self):
        query = """
          query($input: LabelsInput) {
            labels(input: $input) {
                labels {
                    createdAt
                    id
                    label {
                        confidence
                        entity
                        entityType
                        label
                        metadata
                        remove
                        uniqueKey
                    }
                    source {
                        alertHash
                        alertId
                        bot {
                            id
                            image
                            imageHash
                            manifest
                        }
                        chainId
                        id
                    }
                }
                pageInfo {
                    hasNextPage
                    endCursor {
                        pageToken
                    }
                }
            }
          }
          """
        query_variables = vars(self)
        filtered_query_variables = {
            k: v for k, v in query_variables.items() if v is not None}
        return dict(query=query, variables={"input": filtered_query_variables})


class LabelsResponse:
    def __init__(self, dict):
        self.page_info = LabelPageInfo(dict.get('pageInfo')) if dict.get(
            'pageInfo') is not None else None
        self.labels = []
        labels_data = dict.get('labels', [])
        for label_data in labels_data:
            label_dict = label_data.get('label')
            label_dict['id'] = label_data.get('id')
            label_dict['source'] = label_data.get('source')
            label_dict['createdAt'] = label_data.get('createdAt')
            self.labels.append(Label(label_dict))


class LabelPageInfo:
    def __init__(self, dict):
        self.has_next_page = dict.get('hasNextPage')
        self.end_cursor = LabelCursor(dict.get('endCursor')) if dict.get(
            'endCursor') is not None else None


class LabelCursor:
    def __init__(self, dict):
        self.page_token = dict.get('pageToken')

File: sdk/labels.api.ts
import { AxiosInstance } from "axios";
import { EntityType, Label, LabelSource } from "./label";
import { assertExists, getFortaApiHeaders, getFortaApiURL } from "./utils";

export type GetLabels = (query: LabelQueryOptions) => Promise<LabelsResponse>;
export function provideGetLabels(axios: AxiosInstance): GetLabels {
  assertExists(axios, "axios");

  return async function getLabels(
    query: LabelQueryOptions
  ): Promise<LabelsResponse> {
    const response: RawGraphqlLabelResponse = await axios.post(
      getFortaApiURL(),
      getQueryFromLabelOptions(query),
      getFortaApiHeaders()
    );

    if (response.data?.errors?.length)
      throw Error(response.data.errors[0].message);

    const pageInfo = response.data.data.labels.pageInfo;
    const labels: Label[] = [];
    for (const labelData of response.data.data.labels.labels) {
      const { label, id, createdAt, source } = labelData;
      labels.push(
        Label.fromObject({
          ...label,
          metadata: label.metadata ?? {},
          id,
          createdAt,
          source,
        })
      );
    }
    return { labels, pageInfo };
  };
}

export interface LabelQueryOptions {
  entities?: string[];
  labels?: string[];
  sourceIds?: string[];
  entityType?: string;
  state?: boolean;
  createdSince?: number;
  createdBefore?: number;
  first?: number;
  startingCursor?: LabelCursor;
}

export interface LabelCursor {
  pageToken: string;
}

export interface LabelsResponse {
  labels: Label[];
  pageInfo: {
    hasNextPage: boolean;
    endCursor?: {
      pageToken: string;
    };
  };
}

interface RawGraphqlLabelData {
  id: string;
  label: RawGraphqlLabel;
  createdAt: string;
  source: LabelSource;
}

interface RawGraphqlLabel {
  confidence: number;
  entity: string;
  entityType: EntityType;
  label: string;
  metadata: any;
  remove: boolean;
}

interface RawGraphqlLabelResponse {
  data: {
    data: {
      labels: {
        labels: RawGraphqlLabelData[];
        pageInfo: {
          hasNextPage: boolean;
          endCursor?: {
            pageToken: string;
          };
        };
      };
    };
    errors: any;
  };
}

export const getQueryFromLabelOptions = (options: LabelQueryOptions) => {
  return {
    operationName: "fetchLabels",
    query: `
            query fetchLabels(
                $entities: [String],
                $entityType: String,
                $labels: [String],
                $sourceIds: [String],
                $state: Boolean,
                $after: LabelEndCursorInput,
                $first: NonNegativeInt,
                $createdBefore: NonNegativeInt,
                $createdSince: NonNegativeInt
                ) {
                    labels(input:{
                        entities: $entities,
                        entityType: $entityType,
                        labels: $labels,
                        sourceIds: $sourceIds,
                        state: $state,
                        after: $after,
                        first: $first,
                        createdBefore: $createdBefore,
                        createdSince: $createdSince
                    }) {
                        labels {
                            createdAt
                            id
                            label {
                                confidence
                                entity
                                entityType
                                label
                                metadata
                                remove
                            }
                            source {
                              alertHash
                              alertId
                              bot {
                                id
                                image
                                imageHash
                                manifest
                              }
                              chainId
                              id
                            }
                        }
                        pageInfo {
                            hasNextPage
                            endCursor {
                                pageToken
                            }
                        }
                    }
            }
        `,
    variables: {
      entities: options.entities,
      entityType: options.entityType,
      labels: options.labels,
      sourceIds: options.sourceIds,
      state: options.state,
      after: options.startingCursor,
      first: options.first,
      createdBefore: options.createdBefore,
      createdSince: options.createdSince,
    },
  };
};
